### PR TITLE
feat: v0.0.1

### PR DESCRIPTION
**Pull request by mersinvald**
There are some TODOs to be resolved in subsequent versions, but otherwise, it's a working service

- loading pages via Gecko WebDriver
- parsing the data from them
- composing the telegram posts w/ handlebars template
- publishing to telegram

### CODE REVIEW

File: src/config.rs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+use serde::Deserialize;
+use std::path::PathBuf;
+
+#[derive(Deserialize, Debug)]
+pub struct Config {
+    pub update_interval: u64,
+    pub story_db: PathBuf,
+    pub webdriver_host: String,
+    pub webdriver_port: u16,
+    pub telegram_secret: String,
```
Comment (AnthonyMikh) [inline src/config.rs]: Maybe move these fields into substructure?

File: src/main.rs [hunk 2]
```diff
@@ -0,0 +1,135 @@
+#![feature(try_blocks)]
+
+mod config;
+mod loader;
+mod scraper;
+mod state;
+mod tg_bot;
+
+use self::config::Config;
```
Comment (AnthonyMikh) [inline src/main.rs]: `self` can be omitted here

File: src/main.rs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+            if self.state.is_published(&teaser.url)? {
+                continue;
+            }
+
+            let story = self.loader.open(&teaser.url).await?;
+            let story = Story::from_html(&story)?;
+            let formatted = self.format_story(&story, &teaser.url)?;
+
+            self.bot.publish_message(formatted).await?;
+            self.state.set_published(&teaser.url).await?;
```
Comment (AnthonyMikh) [inline src/main.rs]: Shouldn't these two lines be compined into single method?

File: src/main.rs [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        Ok(())
+    }
+
+    fn format_story(&self, story: &Story, url: &str) -> anyhow::Result<String> {
+        let story_content = story
+            .summary
+            .iter()
+            .map(|p| p.telegram_html())
+            .collect::<Vec<String>>()
+            .join("\n\n"); // any way to do that nicely without collecting to Vec<String> first?
```
Comment (AnthonyMikh) [inline src/main.rs]: `itertools::Itertools::join`. Although collecting into `Vec` has the benefit that `.join` can precalculate the needed memory exactly

File: src/scraper.rs [hunk 5]
```diff
@@ -0,0 +1,390 @@
+use anyhow::{anyhow, bail};
+use chrono::{DateTime, FixedOffset};
+use select::document::Document;
```
Comment (AnthonyMikh) [inline src/scraper.rs]: Merge imports maybe?

File: src/scraper.rs [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+        };
+
+        Ok(x)
+    }
+}
+
+impl Side {
+    pub fn emoji(&self) -> &'static str {
+        match *self {
+            Side::Left => "🟦",
```
Comment (AnthonyMikh) [inline src/scraper.rs:71]: Finally, a good use for emojies in code

File: src/scraper.rs [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+
+    pub fn telegram_html(&self) -> String {
+        self.0
+            .children()
+            .map(|n| {
+                Some(n)
+                    .filter(|n| n.is(Name("a")))
+                    .and_then(|a| {
+                        Some(format!(
+                            "<a href=\"{url}\">{text}</a>",
```
Comment (AnthonyMikh) [inline src/scraper.rs]: Use raw string literal maybe?

File: src/scraper.rs [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+            title,
+            summary: paragraphs,
+            articles,
+            datetime,
+        })
+    }
+}
+
+fn normalize_allsides_url(relative_url: &str) -> String {
+    format!("https://www.allsides.com{}", relative_url)
```
Comment (AnthonyMikh) [inline src/scraper.rs]:
```suggestion
    ["https://www.allsides.com{}", relative_url].concat()
```
Also this implies that relative URL begins with `/`. Maybe put that slash here?
Also I saw ALLSIDES_MAINPAGE_URL or something like this. May we get rid of duplication?

File: src/state.rs [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+        Ok(State { stories: db })
+    }
+
+    pub fn is_published(&self, url: &str) -> anyhow::Result<bool> {
+        Ok(self.stories.contains_key(url)?)
+    }
+
+    pub async fn set_published(&mut self, url: &str) -> anyhow::Result<()> {
+        self.stories.insert(url, "true")?;
+        self.stories.flush_async().await?;
```
Comment (AnthonyMikh) [inline src/state.rs:20]: Can we avoid flushing buffers after every insertion?

File: src/tg_bot.rs [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+            channel_id: channel_id.into(),
+            admin_id: admin_id.into(),
+        })
+    }
+
+    pub async fn log_error(&self, err: impl std::fmt::Display) -> anyhow::Result<()> {
+        self.bot
+            .send_message(
+                ChatId::ChannelUsername(self.admin_id.clone()),
+                format!("{}", err),
```
Comment (AnthonyMikh) [inline src/tg_bot.rs]:
```suggestion
                err.to_string(),
```

File: src/tg_bot.rs [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+            .send()
+            .await?;
+        Ok(())
+    }
+
+    pub async fn publish_message(&self, msg: impl std::fmt::Display) -> anyhow::Result<()> {
+        self.bot
+            .send_message(
+                ChatId::ChannelUsername(self.channel_id.clone()),
+                format!("{}", msg),
```
Comment (AnthonyMikh) [inline src/tg_bot.rs]:
```suggestion
                msg.to_string(),
```

File: src/config.rs (hunk 1 above)
Comment (mersinvald) [reply src/config.rs]: With serde(flatten) -- yeah, that's possible

File: src/main.rs (hunk 4 above)
Comment (mersinvald) [reply src/main.rs]: Well, in both cases, performance isn't of concern, but itertools option is cleaner. Thanks!

File: src/state.rs (hunk 9 above)
Comment (mersinvald) [reply src/state.rs:20]: Why?

File: src/tg_bot.rs (hunk 10 above)
Comment (mersinvald) [reply src/tg_bot.rs]: Sure

File: src/config.rs (hunk 1 above)
Comment (mersinvald) [reply src/config.rs]: `envy` breaks on parsing u16 :(

File: src/state.rs (hunk 9 above)
Comment (AnthonyMikh) [reply src/state.rs:20]: I believe that it can unnecessary slow things down

File: src/main.rs [hunk 12]
```diff
@@ -0,0 +1,130 @@
+#![feature(try_blocks)]
```
Comment (AnthonyMikh) [inline src/main.rs]: I don't see them used in code

File: src/main.rs [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+                    "side_story_emoji": article.side.emoji(),
+                    "side_story_title": article.title,
+                    "side_story_url": article.url,
+                    "side_story_source": article.source,
+                    "side_story_content": article_content,
+                })
+            })
+            .collect::<Vec<_>>();
+
+        let side_stories = serde_json::to_value(side_stories)?;
```
Comment (AnthonyMikh) [inline src/main.rs]: Is that just a `let side_stories = serde_json::Value::Array(side_stories);`?

File: src/scraper.rs [hunk 14]
```diff
@@ -0,0 +1,392 @@
...
+        Ok(MainPage { teasers })
+    }
+}
+
+impl<'a> FromHTML<'a> for Story<'a> {
+    fn from_html(html: &'a Document) -> anyhow::Result<Self> {
+        let story = html
+            .find(Attr("id", "content"))
+            .next()
+            .ok_or_else(|| anyhow!("cannot query story body (id=content)"))?;
```
Comment (AnthonyMikh) [inline src/scraper.rs]: Given how often `.ok_or_else(|| anyhow!(<string literal>))` occurs, does make sense to make an extension trait for `Result` to provide a shortcut?

File: src/state.rs (hunk 9 above)
Comment (mersinvald) [reply src/state.rs:20]: It needs to process a few posts per day on average. It's far from a performance-bound task, but ensuring that posts are not duplicated is important

File: src/main.rs (hunk 13 above)
Comment (mersinvald) [reply src/main.rs]: It is, yeah

File: src/main.rs (hunk 12 above)
Comment (mersinvald) [reply src/main.rs]: Yeah, I used them initially, but then replaced with regular iterators

File: src/scraper.rs (hunk 14 above)
Comment (mersinvald) [reply src/scraper.rs]: Since those aren't actually using formatting, yeah, good idea

File: src/main.rs (hunk 12 above)
Comment (AnthonyMikh) [reply src/main.rs]: Then remove it maybe?

**Pull request merged**